frontend/src/globals/unity-indexeddb-worker.ts

```diff
@@ -0,0 +1,359 @@
+	 * when set requests are made. It is used to perform early rejects on get
+	 * requests we know will fail (because there is no key in the store),
+	 * avoiding expensive database transactions. */
+	index: any;
+
+	createIndexedDbCache() {
+		const request = indexedDB.open('3DRepoCacheDb', 1);
+
+		request.onerror = () => {
+			console.error('Unable to open IndexedDb - Model Caching will be Disabled.');
+			this.sendIndexedDbUpdated({
+				state: 'Disabled',
+			});
+		};
+
+		// When onupgradeneeded completes successfully, onsuccess will be called
+		request.onupgradeneeded = (event: any) => {
+			const db = event.target.result;
+			db.createObjectStore(this.objectStoreName, {}); // The database uses a simple key-pair assocation, where the key is passed explicitly
+		};
+
+		request.onsuccess = (event: any) => {
+			this.db = event.target.result;
+
+			// Populate our index
+			const transaction = this.db.transaction(this.objectStoreName, 'readonly');
+			const objectStore = transaction.objectStore(this.objectStoreName);
+			const req = objectStore.getAllKeys();
+			req.onsuccess = () => {
+				this.index = {};
+				req.result.forEach((x) => this.index[x] = 1); // Give any value here - we will only check the existence of the key (property)
+				this.sendIndexedDbUpdated({
+					state: 'Open',
+				});
+			};
+
+			this.db.onerror = (error: any) => {
+				console.error(`IndexedDbCache error has bubbled to db. This should not happen as all errors should be handled within request error handlers: ${error}`);
+			};
+			this.db.onversionchange = () => {
+				this.db.close();
+			};
+		};
+	}
+
+	recreateIndexedDb() {
+		// This method is called when a transaction has raised a state exception.
+
+		// We can try and reconnect. In the meantime signal the viewer to hold
+		// all additional transactions.
+		this.sendIndexedDbUpdated({
+			state: 'Pending',
+		});
+
+		this.db = undefined;
+		this.createIndexedDbCache();
+	}
+
+	createGetTransaction(id: number, key: string) {
+		// Before anything else, check if the key is in our local memory cache
+		const record = this.memoryCache[key];
+		if (record !== undefined) {
+			this.sendGetTransactionComplete({
+				id,
+				size: record.size,
+				result: record,
+			});
+		} else if (!this.index.hasOwnProperty(key)) {
+			// We know the database doesnt have the key, so return immediately again
+			this.sendGetTransactionComplete({
+				id,
+				size: -1,
+			});
+		} else {
+			// We believe the key is in the database, so go ahead and request it
+			try {
+				const readTransaction = this.db.transaction(this.objectStoreName, 'readonly');
+				const objectStore = readTransaction.objectStore(this.objectStoreName);
+				const request = objectStore.get(key);
+
+				request.onsuccess = () => {
+					if (request.result === undefined) {
+						this.sendGetTransactionComplete({ // Where a key is not found, we want to handle this outside the error chain
```

Comment:
so it's possible for us to think the key is in the db but it was not there after all?

Reply:
Yes! It should be incredibly rare, but it is possible... 
One example is if the user clears the database out from under the tab, either manually or because an onVersionChanged event rebuilds it. When the database is reopened the dictionary should be rebuilt, but anything that happens between then could use out of date data.
Another possibility is that the commit transaction fails after one or more individual set operation callbacks complete, which will roll back all the changes in that transaction.